### ISSUE TITLE
With kube-proxy replacement (some) pods on new nodes (partially?) missing connectivity until Cilium agent is restarted

### ISSUE THREAD

**Issue by choppedpork** [cilium/cilium#28764]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

I'm currently working on rolling Cilium out to a few dozen k8s clusters. I've done a successful PoC on one last year with version 1.11 and have now upgraded to 1.14.2 and trying to tweak the feature set (some combinations were not possible in 1.11). The main two items I want to enable is encryption for pod-to-pod traffic as well as the kube-proxy replacement.

In the 1.11 setup this was not possible as kube-proxy replacement did not (and still does not) work with IPsec. I've successfully switched to Wireguard and things have been running great until I've tried enabling kube-proxy replacement too.

Whenever a new k8s node comes up, some (but not all) pods scheduled on it struggle with connectivity - both ingress and egress. I've initially noticed this as some pods would get into a CrashLoopBackOff state and upon inspection I've seen many different forms of timeouts in the logs (depending on how the workload initialises):
- dns resolution
- kubernetes API calls
- "world" egress (eg. EC2 metadata API)
- affected pods also can't receive traffic

Once the cilium agent on the node is restarted all affected pods start working correctly - no need to restart or reschedule, so long as they retry everything is fine. When new pods are scheduled on an affected node they will also _sometimes_ get into this state - I've been using a `nodeSelector` to target a known bad node to run a very simple http-echo service as well as busybox, the echo service worked only once out of a few attempts, the busybox pod had quite inconsistent bad behaviour (I was not able to access the control plane or egress but I _was_ able to resolve DNS entries and egress to pods and services in the cluster - but not "world").

I've verified that the behaviour is the same with and without Wireguard - whenever I redeploy with `--set kubeProxyReplacement=true` and rotate nodes (+ remove kube-proxy) the issue reappears.

### Cilium Version

1.14.2

### Kernel Version

5.10.192-183.736.amzn2.x86_64

### Kubernetes Version

v1.25.12-eks-2d98532

### Anything else?

I've not been able to find any relevant logs - it seems like according to the cilium agent on the node everything is actually fine. I've ran a `cilium connectivity test --single-node` against a node in a bad state and the results where as expected:

```
connectivity test failed: timeout reached waiting for lookup for kubernetes.default from pod cilium-test/client-5887967774-r46b9 to succeed (last error: context deadline exceeded)
```

One thing I've done is `tcpdump -i cilium_wg0` on the node, filtering by the IP of affected pods. The result was quite interesting - all the traffic that times out was not present on that interface at all. I've mentioned above that I brought up and exec'd into a busybox pod, I could see all the "successful" traffic in the tcpdump output - dns requests, requests to in-cluster services - but not the traffic that was not working (the k8s control plane at 172.20.0.1 or world). I've not re-done this particular test with Wireguard disabled.

Here's my sanitised values file (NB I've removed the custom image locations - they are simply official images mirrored into our ECR):

```yaml
bandwidthManager:
  enabled: false
egressMasqueradeInterfaces: eth+
enableIPv4Masquerade: false
encryption:
  enabled: true
  type: wireguard
eni:
  awsReleaseExcessIPs: true
  ec2APIEndpoint: ec2.region.api.aws
  enabled: true
  eniTags:
    bai:cluster: foo
    bai:environment: dev
    bai:no-kube2iam: "true"
    bai:team: sre
  iamRole: arn:aws:iam::foo:role/bar-cilium-operator-role
  subnetTagsFilter: kubernetes.io/cluster/bar=owned
  updateEC2AdapterLimitViaAPI: true
hostFirewall:
  enabled: false
hubble:
  metrics:
    enabled:
    - dns
    - drop
    - tcp
    - flow
    - icmp
    - httpV2
    serviceMonitor:
      enabled: true
  relay:
    enabled: true
  tls:
    auto:
      certValidityDuration: 1095
      enabled: true
      method: cronJob
      schedule: 0 0 1 */4 *
  ui:
    enabled: true
    ingress:
      annotations:
        kubernetes.io/ingress.class: nginx
      enabled: true
      hosts:
      - hubble.foo
installNoConntrackIptablesRules: true
ipam:
  mode: eni
k8sServiceHost: baz.gr7.region.eks.amazonaws.com
kubeProxyReplacement: true
loadBalancer:
  algorithm: maglev
localRedirectPolicy: true
logOptions:
  format: json
nodePort:
  enabled: false
operator:
  prometheus:
    enabled: true
    serviceMonitor:
      enabled: true
  rollOutPods: true
priorityClassName: system-node-critical
prometheus:
  enabled: true
  serviceMonitor:
    enabled: true
rollOutCiliumPods: true
routingMode: native
serviceAccounts:
  operator:
    annotations:
      eks.amazonaws.com/audience: sts.region.amazonaws.com
socketLB:
  hostNamespaceOnly: false
```

You'll notice some of the options are set to defaults - these were added as feature toggles to try things out. The `socketLB` option is something we need enabled on some of our clusters (which use Istio).

What would be a good next thing to check? I'm running on the assumption that this is a configuration issue - I believe I do have all the required settings in place for EKS though.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by choppedpork** [comment]
I've not been able to fully confirm this yet (only did a spot check and double node rotation in a single cluster, will try more clusters tomorrow) but it looks like enabling `eni.awsEnablePrefixDelegation` resolves the issue.

**Comment by choppedpork** [comment]
Not sure if this muddies the waters further but while enabling prefix delegation has been working fine on a few clusters (we're in the process of deploying this configuration to more as I speak), switching `socketLB.hostNamespaceOnly` to `true` - on a cluster where it's needed due to the use of Istio - has caused the issue to occur again. Switching off the kube proxy replacement allowed for a perfectly working install though.

**Comment by msolters** [comment]
> One thing I've done is tcpdump -i cilium_wg0 on the node, filtering by the IP of affected pods. The result was quite interesting - all the traffic that times out was not present on that interface at all. I've mentioned above that I brought up and exec'd into a busybox pod, I could see all the "successful" traffic in the tcpdump output - dns requests, requests to in-cluster services - but not the traffic that was not working (the k8s control plane at 172.20.0.1 or world). I've not re-done this particular test with Wireguard disabled.

In those situations where these symptoms occur, what does the content of `cilium debuginfo --output json | jq '.encryption'` show?

As [discussed here](https://docs.cilium.io/en/latest/security/network/encryption-wireguard/#troubleshooting), are the kpods with degraded connectivity listed under the corresponding `peers[].allowed-ips` entry?

On the problematic node, are remote kpods missing from `peers[].allowed-ips`?

On non-problematic nodes, are remote kpods on the problematic node missing from `peers[].allowed-ips`?

Basically I'm just wondering if perhaps some kpods just stop being enumerated in `allowed-ips` somewhere and restarting cilium-agent fixes it.

However,

> I've verified that the behaviour is the same with and without Wireguard - whenever I redeploy with --set kubeProxyReplacement=true and rotate nodes (+ remove kube-proxy) the issue reappears.

This makes me think the issue is deeper than Wireguard and `allowed-ips`.
